apps/dashboard/tests/page-models/create-workflow-sidebar.ts

```diff
@@ -0,0 +1,70 @@
+import { expect, type Page } from '@playwright/test';
+
+export class CreateWorkflowSidebar {
+  constructor(private page: Page) {}
+
+  async createBtnClick({ awaitResponse = false } = {}): Promise<void> {
+    const createWorkflowBtn = await this.page.getByRole('button', { name: 'Create workflow' });
+    await createWorkflowBtn.click();
+
+    if (awaitResponse) {
+      await this.page.waitForResponse(
+        (resp) => resp.url().includes('/v2/workflows') && resp.request().method() === 'POST' && resp.status() === 201
+      );
+    }
+  }
+
+  async expectNameValidationError(): Promise<void> {
+    await expect(await this.page.getByText('Name is required')).toBeVisible();
```

Comment:
Page object models are usually selectors and request helpers that keep the e2e code dry. To me, expectations belong to the test. Moreover, E2E tests have to be readable and self-explanatory. Dryness is not the most critical requirement.

So, I suggest adding all the expectations in the e2e file and using the page object models for selectors, event listeners and waiting logic. This ensures that page object models are highly reusable and can work with any expectation the e2e will require.

Most importantly, when the e2e is read by the future maintainer, they don't have to jump into every function to see what it does.

Reply:
I agree.